src/Platform/Microsoft.Testing.Platform/Services/TestApplicationResult.cs

```diff
@@ -13,19 +13,40 @@
 
 namespace Microsoft.Testing.Platform.Services;
 
-internal sealed class TestApplicationResult(
-    IOutputDevice outputService,
-    ITestApplicationCancellationTokenSource testApplicationCancellationTokenSource,
-    ICommandLineOptions commandLineOptions,
-    IEnvironment environment) : ITestApplicationProcessExitCode, IOutputDeviceDataProducer
+internal sealed class TestApplicationResult : ITestApplicationProcessExitCode, IOutputDeviceDataProducer
 {
-    private readonly IOutputDevice _outputService = outputService;
-    private readonly ITestApplicationCancellationTokenSource _testApplicationCancellationTokenSource = testApplicationCancellationTokenSource;
-    private readonly ICommandLineOptions _commandLineOptions = commandLineOptions;
-    private readonly IEnvironment _environment = environment;
+    private readonly IOutputDevice _outputService;
+    private readonly ITestApplicationCancellationTokenSource _testApplicationCancellationTokenSource;
+    private readonly ICommandLineOptions _commandLineOptions;
+    private readonly IEnvironment _environment;
     private readonly List<TestNode> _failedTests = [];
     private int _totalRanTests;
     private bool _testAdapterTestSessionFailure;
+    private bool _testExecutionStopped;
+
+    public TestApplicationResult(
+        IOutputDevice outputService,
+        ITestApplicationCancellationTokenSource testApplicationCancellationTokenSource,
+        ICommandLineOptions commandLineOptions,
+        IEnvironment environment,
+        PoliciesService policiesService)
+    {
+        _outputService = outputService;
+        _testApplicationCancellationTokenSource = testApplicationCancellationTokenSource;
+        _commandLineOptions = commandLineOptions;
+        _environment = environment;
+        policiesService.RegisterOnStopTestExecution(
+            async cancellationToken =>
+            {
+                // How to make the message more clear that it's due to --max-failed-tests?
+                // The way we are currently abstracting is too generic that any one could use IStopTestExecutionCapability for whatever other reason.
+                // So, we can't here "hardcode" the message to be specific to --max-failed-tests.
+                // Do we want to introduce a "reason" parameter in IStopTestExecutionCapability.StopTestExecutionAsync?
+                // TODO: Localize
+                await _outputService.DisplayAsync(this, new TextOutputDeviceData("Test session was stopped."));
```

Comment:
I think it's not duty of this class. Here we should only use this call back to return the new exit code.
The `IPlatformOutputDevice` should register and decide "how to show" this event.